analyses/malawi/notebooks/mwi_ecmwf_monthly_skill_false_alarms.md

```diff
@@ -0,0 +1,308 @@
+df_haz=df_mvac[["District","TA Name 2018 Cencus","Year MVAC Started Analyzing TA"]+hazard_cols].dropna(how="all")
+df_haz.rename(columns={"TA Name 2018 Cencus":"TA","Year MVAC Started Analyzing TA":"start_analysis"},inplace=True)
+#switch year and hazard, in order to apply wide_to_long
+for i in hazard_cols:
+    year=i[:4]
+    df_haz[f"Hazard {year}"]=df_haz[i]
+#get one row per TA and year with the hazard
+df_hazl=pd.wide_to_long(df_haz,"Hazard",i=["District","TA"],j="year",sep=" ")[["start_analysis","Hazard"]].reset_index()
+#compute whether a dry spell occured
+df_hazl['dry_spell'] = df_hazl.apply(lambda x: ab_ify(x),axis=1)
+gdf_adm2=gpd.read_file(adm2_bound_path)
+df_hazl=df_hazl.merge(gdf_adm2[["ADM2_EN","ADM1_EN"]],how="left",left_on="District",right_on="ADM2_EN")
+df_mvac_year=df_hazl.groupby(["year","ADM1_EN"])["dry_spell"].agg(ds_sum="sum",ds_count="count").reset_index()
+#compute percentage of analyzed TAs experiencinga dry spell per year
+df_mvac_year["perc"]=df_mvac_year.ds_sum/df_mvac_year.ds_count*100
+# MVAC's year is the year during which thre crops were harvested, i.e. the end of the rainy season
+#While for the obs dry spells from CHIRPS we use season_approx as the year the rainy season started --> add that variable to the dataset
+df_mvac_year.rename(columns={"year":"season_approx_mvac"},inplace=True)
+df_mvac_year["season_approx"]=df_mvac_year.season_approx_mvac-1
+```
+
+```python
+df_mvac_year[df_mvac_year.ADM1_EN=="Southern"].head()
+```
+
+```python
+#plot the percentages across years (ugly just for inspection)
+df_mvac_year[df_mvac_year.ADM1_EN=="Southern"].plot(x="season_approx",y="perc",kind="bar",figsize=(5,2),legend=False,title="percentage of area experiencing a dry spell")
+```
+
+```python
+#start month of the rainy season
+start_rainy_seas=10
+#season approx indicates the year during which the rainy season started
+#this is done because it can start during one year and continue the next calendar year
+#we therefore prefer to group by rainy season instead of by calendar year
+df_forobs["season_approx"]=np.where(df_forobs.date_month.dt.month>=start_rainy_seas,df_forobs.date_month.dt.year,df_forobs.date_month.dt.year-1)
+```
+
+```python
+#group to rainy season, as mvac is yearly data
+df_forobs_seas=df_forobs.groupby(["season_approx","leadtime","ADM1_EN"],as_index=False).sum()
+```
+
+```python
+df_forobs_seas=df_forobs_seas.merge(df_mvac_year[["season_approx","ADM1_EN","perc"]],how="left",on=["season_approx","ADM1_EN"])
+df_forobs_seas.rename(columns={"perc":"perc_ds_mvac"},inplace=True)
+```
+
+```python
+#note: a neater version of computing the rp period will be implemented in utils_general/statistics.py soon
+from scipy.interpolate import interp1d
+def get_return_period_function(df,var):
+    df_rp=df.sort_values(by=var,ascending=False)
+    n=len(df)
+    #rank most extreme occurrences
+    df_rp['rank'] = np.arange(n) + 1
+    df_rp['exceedance_probability'] = df_rp['rank'] / (n+1)
```

Comment:
I don't understand how exceedance probability is computed. Is it the probability that a value exceeds the one ranked? Doesn't arrange(n) sort in increasing order?  Why are we adding 1's?

Reply:
Thanks for asking! 

> Is it the probability that a value exceeds the one ranked?
It is the rank divided by the total number of entries. The highest rank indicates the highest value, so the exceedance probability of this will be the lowest. Does that clarify more? 

> Doesn't arrange(n) sort in increasing order?
arange(n) creates an array with numbers from 0 to n-1. Sort_values above is used to sort the values. You are right that by default it sorts in increasing order, but that is why the argument `ascending=False` is added. 

> Why are we adding 1's?
Cause else the lowest exceedence probability will be 0, which will cause a zero division with `1 / df_rp['exceedance_probability']` We could also add an epsilon instead, but then this division will become very large for the lowest exceedence probability while with this method it is closer to n, which makes more sense.

Reply:
Also, this is an empirical fitting method, one could also use an analytical method (though in this case with this distribution that doesn't work, can explain more if you want). We implemented these methods now also in `utils_general` so changed the code to use that.